File: Data Structures & Algorithms/evaluate-reverse-polish-notation/submission-4.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []

        for tok in tokens:
            
            if tok == "+":
                cur = int(stack.pop()) + int(stack.pop())
                stack.append(cur)
            elif tok == "-":
                B = int(stack.pop())
                A = int(stack.pop())
                cur = A - B
                stack.append(cur)
            elif tok == "*":
                cur = int(stack.pop()) * int(stack.pop())
                stack.append(cur)
            elif tok == "/":
                B = int(stack.pop())
                A = int(stack.pop())
                cur = A / B
                stack.append(cur)
            else:
                stack.append(int(tok))

        return int(stack.pop())

            

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-4.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {

        int res = 0;
        stack<vector<int>> s;
        vector<int> rect;
        int start;

        for(int i = 0; i < heights.size(); ++i){
            start = i;
            while(!s.empty() && s.top()[1] > heights[i]){
                rect = s.top();
                s.pop();
                res = max(res, (i - rect[0]) * rect[1]);
                start = rect[0];
            }
            s.push({start, heights[i]});
        }

        while(!s.empty()){
            rect = s.top();
            s.pop();
            res = max(res, ((int)heights.size() - rect[0]) * rect[1]);
            start = rect[0];
        }

        return res;
    }
};
